### PR TITLE
PTDF factorize

### PR DESCRIPTION
**Pull request by bknueven**
Add `VirtualPTDFMatrix` object, which calculates PTDF rows and solves for flows on-the-fly. DCOPF and unit commitment are changed to use this by default.

Performance benefit (as measured for DCOPF) for `pglib_opf_case1354_pegase`, using Pyomo 5.7.1 and `gurobi_persistent`:
|    |      master      |  PR |
|----------:|:-------------:|:------:|
| PTDF precalculation  |   48.80    |  0.12 |
| PTDF solve total time |  59.19   |  6.09  |
| B-theta solve time |     15.23   |  15.33  |

Major improvement driven by using `scipy.sparse.linalg.splu`, per @michaelbynum and @carldlaird, and re-organizing various code around that.

Minor changes:
1. Minor refactoring existing code to support this
1. Deprecates PTDF serialization/deserialization as it's not really necessary anymore

### CODE REVIEW

File: egret/model_library/transmission/tx_calc.py [hunk 1]
```diff
@@ -547,6 +565,89 @@ def _calculate_pfl_constant(branches,buses,index_set_branch,base_point=BasePoint
...
+                branch_idx = mapping_branch_to_idx[l]
+                row.append(idx)
+                col.append(branch_idx)
+                data.append(val)
+            else:
+                ## TODO: do we need enforce this requirement?
+                raise Exception("Interface {} has line {} with line_orientation {} "
+                        "not in [-1,0,1].".format(i_n, l, val))
+    I = sp.coo_matrix((data,(row,col)),shape=(_len_interface,_len_branch))
+    return I.tocsr()
```
Comment (michaelbynum) [inline egret/model_library/transmission/tx_calc.py:649]: Is there a need to convert this to csr? In my experience, coo matrix dot products are just as efficient, if not more so.

**Review by michaelbynum — APPROVED** [review]
Nice! Overall, this looks good. The public methods in the `VirtualPTDFMatrix` could use some documentation, but who am I to complain about documentation...

File: egret/model_library/transmission/tx_calc.py (hunk 1 above)
Comment (bknueven) [reply egret/model_library/transmission/tx_calc.py]: No, I don't think so. We don't do much with this matrix, so I'll leave it as a `coo_matrix`.

**Comment by bknueven** [comment]
@michaelbynum I added some docstrings, which was a good thing to do because it forced me to make a few minor changes to make the interface smoother.

**Comment by michaelbynum** [comment]
Nice! Can we merge this now?

**Comment by bknueven** [comment]
Sure, I think it's done.

**Pull request merged**